pkg/operator/k8sutil/cmdreporter/cmdreporter.go

```diff
@@ -0,0 +1,351 @@
+) (*CmdReporter, error) {
+	if clientset == nil || ownerRef == nil {
+		return nil, fmt.Errorf("clientset [%+v] and owner reference [%+v] must be specified", clientset, ownerRef)
+	}
+	if appName == "" || jobName == "" || jobNamespace == "" {
+		return nil, fmt.Errorf("app name [%s], job name [%s], and job namespace [%s] must be specified", appName, jobName, jobNamespace)
+	}
+	// at least one command must be set, and it cannot be an empty string
+	if len(cmd) == 0 || cmd[0] == "" {
+		return nil, fmt.Errorf("command [%+v] must be specified", cmd)
+	}
+	if rookImage == "" || runImage == "" {
+		return nil, fmt.Errorf("Rook image [%s] and run image [%s] must be specified", rookImage, runImage)
+	}
+	cfg := &cmdReporterCfg{
+		clientset:    clientset,
+		ownerRef:     ownerRef,
+		jobName:      jobName,
+		jobNamespace: jobNamespace,
+		cmd:          cmd,
+		args:         args,
+		rookImage:    rookImage,
+		runImage:     runImage,
+	}
+
+	job, err := cfg.initJobSpec()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes job spec for CmdReporter job %s. %+v", jobName, err)
+	}
+	return &CmdReporter{
+		clientset: clientset,
+		job:       job,
+	}, nil
+}
+
+// Job returns a pointer to the basic, filled-out Kubernetes job which will run the CmdReporter. The
+// operator may add additional information to this spec, such as labels, environment variables,
+// volumes, volume mounts, etc. before the CmdReporter is run.
+func (cr *CmdReporter) Job() *batch.Job {
+	return cr.job
+}
+
+// Run runs the Kubernetes job and waits for the output ConfigMap. It returns the stdout, stderr,
+// and retcode of the command as long as the image ran it, even if the retcode is nonzero (failure).
+// An error is reported only if the command was not run to completion successfully. When this
+// returns, the ConfigMap is cleaned up (destroyed).
+func (cr *CmdReporter) Run(timeout time.Duration) (stdout, stderr string, retcode int, retErr error) {
+	jobName := cr.job.Name
+	namespace := cr.job.Namespace
+	errMsg := fmt.Sprintf("failed to run CmdReporter %s successfully.", jobName)
+
+	// the configmap MUST be deleted, because we will wait on its presence to determine when the
+	// job is done running
+	delOpts := &k8sutil.DeleteOptions{}
+	delOpts.Wait = true
+	delOpts.ErrorOnTimeout = true
+	// configmap's name will be the same as the app
+	err := k8sutil.DeleteConfigMap(cr.clientset, jobName, namespace, delOpts)
+	if err != nil {
+		return "", "", -1, fmt.Errorf("%s. failed to delete existing results ConfigMap %s. %+v", errMsg, jobName, retErr)
+	}
+
+	if err := k8sutil.RunReplaceableJob(cr.clientset, cr.job, true); err != nil {
+		return "", "", -1, fmt.Errorf("%s. failed to run job. %+v", errMsg, err)
+	}
+
+	listOpts := metav1.ListOptions{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "ConfigMap",
+		},
+		FieldSelector: fmt.Sprintf("metadata.name=%s", jobName),
+	}
+
+	list, err := cr.clientset.CoreV1().ConfigMaps(namespace).List(listOpts)
```

Comment:
Yeah... this was following the pattern that the OSDs use to watch for ConfigMap changes. But now that you say it, this seems a little overengineered for the task at hand.

Reply:
Keeping this as a List+Watch, because this is less load on the API server. Pulled some functionality out into functions so this isn't using `break`/`goto` with labels. And I changed the behavior after list that if there are items listed, we avoid a race condition where the job has already run and created the configmap before list returns stuff. Also clarified that one notable reason for watch channel hangups is that k8s has cleared its change history.

https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes
> A given Kubernetes server will only preserve a historical list of changes for a limited time. Clusters using etcd3 preserve changes in the last 5 minutes by default. When the requested watch operations fail because the historical version of that resource is not available, clients must handle the case by recognizing the status code 410 Gone, clearing their local cache, performing a list operation, and starting the watch from the resourceVersion returned by that new list operation.